flutter/assets/directory_asset_bundle.cc

```diff
@@ -0,0 +1,55 @@
+// Copyright 2016 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#include "flutter/assets/directory_asset_bundle.h"
+
+#include <fcntl.h>
+#include <unistd.h>
+
+#include <utility>
+
+#include "glue/data_pipe_utils.h"
+#include "lib/ftl/files/path.h"
+
+namespace blink {
+
+void DirectoryAssetBundle::Create(
+    mojo::InterfaceRequest<mojo::asset_bundle::AssetBundle> request,
+    std::string directory,
+    ftl::RefPtr<ftl::TaskRunner> task_runner) {
+  new DirectoryAssetBundle(std::move(request), std::move(directory),
+                           std::move(task_runner));
+}
+
+void DirectoryAssetBundle::GetAsStream(
+    const mojo::String& asset_name,
+    const mojo::Callback<void(mojo::ScopedDataPipeConsumerHandle)>& callback) {
+  mojo::DataPipe pipe;
+  callback.Run(std::move(pipe.consumer_handle));
+
+  std::string asset_path =
+      files::SimplifyPath(directory_ + "/" + asset_name.get());
+  if (asset_path.find(directory_) != 0u) {
+    FTL_LOG(ERROR) << "Asset name '" << asset_name
+                   << "' attempted to traverse outside asset bundle.";
+    return;
+  }
+
+  // TODO(abarth): Consider moving the |open| call to task_runner_.
+  ftl::UniqueFD fd(open(asset_path.c_str(), O_RDONLY));
```

Comment:
Also, a check for non-negative integer so some of the later steps can be avoided in case of failure.